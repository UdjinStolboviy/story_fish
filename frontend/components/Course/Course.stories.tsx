import { expect } from "@storybook/jest";
import { screen } from "@storybook/testing-library";
import { ComponentStoryObj, ComponentMeta } from "@storybook/react";

import { Course } from "./Course";

export default {
  title: "Content/Course",
  component: Course,
} as ComponentMeta<typeof Course>;

export const BasicCourse: ComponentStoryObj<typeof Course> = {
  play: async () => {
    await expect(screen.getByRole("heading")).toBeInTheDocument();
    await expect(screen.getByRole("img")).toBeInTheDocument();
    await expect(screen.getByRole("link")).toBeInTheDocument();
  },
  args: {
    header: "Hands-On React. Build advanced React JS Frontend with expert",
    link: "/hands-on-reactjs",
    imageProps: {
      width: 1368,
      height: 770,
      alt: "Logo for Hands-On React. Build advanced React JS Frontend with expert",
      src: "/covers/hands-on_reactjs_cover.png",
    },
    children: (
      <>
        React is the most popular library for building frontend web
        applications. Step-by-step by diving into all the basics, I&apos;ll
        introduce you to advanced concepts as well. We&apos;ll build the
        minesweeper application from scratch We&apos;ll build the minesweeper
        application from scratch:
        <ul>
          <li>setup of the development environment</li>
          <li>configuration of the React JS app</li>
          <li>basic algorithms of Minesweeper</li>
        </ul>
      </>
    ),
  },
  argTypes: {
    children: {
      control: false,
    },
  },
};
