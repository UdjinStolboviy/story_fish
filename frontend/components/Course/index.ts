export { Course, Wrapper, Courses } from "./Course";
