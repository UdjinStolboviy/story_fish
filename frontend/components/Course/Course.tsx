import { FC } from "react";
import Link from "next/link";
import Image, { ImageProps } from "next/image";
import styled from "@emotion/styled";

import { Course as CourseType } from "@/types";

import { boxShadow, borderRadius } from "@/components/styles";
import { StyledLink } from "@/components/StyledLink";

const Section = styled.section`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  padding: 2vmin;
  background: ${({ theme }) => theme.background};
  color: ${({ theme }) => theme.font.regular};
  ${borderRadius};
  ${({ theme }) =>
    boxShadow(theme.components.shadow1, theme.components.shadow2)};
`;

const CourseLink = styled(StyledLink)`
  display: flex;
  width: 94vw;
  @media (min-width: 900px) {
    width: 46vw;
  }
`;

export type Props = {
  /** Header string */
  header: string;
  /** Link address */
  link: string;
  /** Image props */
  imageProps: ImageProps;
};

export const Course: FC<Props> = ({ children, header, link, imageProps }) => (
  <Link href={link} passHref>
    <CourseLink>
      <Section>
        <h2>{header}</h2>
        <Image {...imageProps} alt={header} />
        {children}
      </Section>
    </CourseLink>
  </Link>
);

export const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  gap: 2vw;
  margin: 2vh 1vw;
`;

export const Courses: FC<{ courses: CourseType[]; strapi_url: string }> = ({
  courses,
  strapi_url,
}) => (
  <Wrapper>
    {courses?.map(
      ({
        id,
        attributes: {
          header,
          subtitle,
          publishedAt,
          cover: {
            data: {
              attributes: {
                formats: {
                  medium: { url, width, height },
                },
              },
            },
          },
        },
      }) => (
        <Course
          key={id}
          header={header}
          link={`/course/${id}`}
          imageProps={{
            width,
            height,
            alt: `Cover for ${header}`,
            src: `${strapi_url}${url}`,
          }}
        >
          <h3>{subtitle}</h3>
          <time dateTime={publishedAt}>
            {new Date(publishedAt).toDateString()}
          </time>
        </Course>
      )
    )}
  </Wrapper>
);
