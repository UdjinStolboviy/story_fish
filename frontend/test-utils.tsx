import { FC, ReactElement } from "react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@emotion/react";

import { rootReducer, RootState } from "./store";
import { Themes } from "./styles/themes";
import { Layout } from "./components/Layout";

type Options = { preloadedState?: RootState } & RenderOptions;

const customRender = (
  ui: ReactElement,
  { preloadedState, ...options }: Options = {}
) => {
  const store = configureStore({ reducer: rootReducer, preloadedState });

  const Wrapper: FC = ({ children }) => (
    <Provider store={store}>
      <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
    </Provider>
  );
  return render(ui, { wrapper: Wrapper, ...options });
};

const pageRender = (
  ui: ReactElement,
  { preloadedState, ...options }: Options = {}
) => {
  const store = configureStore({ reducer: rootReducer, preloadedState });

  const Wrapper: FC = ({ children }) => (
    <Provider store={store}>
      <Layout>{children}</Layout>
    </Provider>
  );
  return render(ui, { wrapper: Wrapper, ...options });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render, pageRender };
