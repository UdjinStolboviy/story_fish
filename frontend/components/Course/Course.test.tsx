import { render } from "@/test-utils";

import { Course } from "./Course";

describe("Course test cases", () => {
  it("Course render check", () => {
    const { asFragment } = render(
      <Course
        header="Hands-On React. Build advanced React JS Frontend with expert"
        link="/hands-on-reactjs"
        imageProps={{
          width: 1368,
          height: 770,
          alt: "Logo for Hands-On React",
          src: "/hands-on_reactjs_cover.png",
        }}
      >
        <>
          React is the most popular library for building frontend web
          applications. Step-by-step by diving into all the basics, I&apos;ll
          introduce you to advanced concepts as well. We&apos;ll build the
          minesweeper application from scratch We&apos;ll build the minesweeper
          application from scratch:
          <ul>
            <li>setup of the development environment</li>
            <li>configuration of the React JS app</li>
            <li>basic algorithms of Minesweeper</li>
          </ul>
        </>
      </Course>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
