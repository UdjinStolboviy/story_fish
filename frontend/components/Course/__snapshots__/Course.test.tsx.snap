// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course test cases Course render check 1`] = `
<DocumentFragment>
  <a
    class="css-1sbvujj"
    href="/hands-on-reactjs"
  >
    <section
      class="css-plxdpn"
    >
      <h2>
        Hands-On React. Build advanced React JS Frontend with expert
      </h2>
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%271368%27%20height=%27770%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="Hands-On React. Build advanced React JS Frontend with expert"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
      React is the most popular library for building frontend web applications. Step-by-step by diving into all the basics, I'll introduce you to advanced concepts as well. We'll build the minesweeper application from scratch We'll build the minesweeper application from scratch:
      <ul>
        <li>
          setup of the development environment
        </li>
        <li>
          configuration of the React JS app
        </li>
        <li>
          basic algorithms of Minesweeper
        </li>
      </ul>
    </section>
  </a>
</DocumentFragment>
`;
